File: src/main/java/com/jcloisterzone/feature/GamblersLuckShield.java
package com.jcloisterzone.feature;

import com.jcloisterzone.board.Location;
import com.jcloisterzone.board.Position;
import com.jcloisterzone.board.Rotation;
import com.jcloisterzone.board.pointer.FeaturePointer;
import com.jcloisterzone.game.capability.GamblersLuckCapability.GamblersLuckShieldToken;
import com.jcloisterzone.game.state.GameState;

import io.vavr.collection.List;

public class GamblersLuckShield extends TileFeature implements Structure {

    private final GamblersLuckShieldToken shieldToken;

    public GamblersLuckShield(Location loc) {
        this(List.of(new FeaturePointer(Position.ZERO, GamblersLuckShield.class, loc)), null);
    }

    public GamblersLuckShield(List<FeaturePointer> places, GamblersLuckShieldToken shieldToken) {
        super(places);
        this.shieldToken = shieldToken;
    }

    @Override
    public GamblersLuckShield placeOnBoard(Position pos, Rotation rot) {
        return new GamblersLuckShield(placeOnBoardPlaces(pos, rot), shieldToken);
    }

    public GamblersLuckShield setShieldToken(GamblersLuckShieldToken shieldToken) {
        return new GamblersLuckShield(places, shieldToken);
    }

    public GamblersLuckShieldToken getShieldToken() {
        return shieldToken;
    }

    public static String name() {
        return "GamblersLuckShield";
    }
}

File: src/main/ts/com/jcloisterzone/feature/GamblersLuckShield.ts
import { List } from "../../../io/vavr/SeqTypes.js";
import { Location } from "../board/Location.js";
import { Position } from "../board/Position.js";
import type { Rotation } from "../board/Rotation.js";
import { FeaturePointer } from "../board/pointer/FeaturePointer.js";
import type { GamblersLuckCapability } from "../game/capability/GamblersLuckCapability.js";
import { TileFeature } from "./TileFeature.js";

type GamblersLuckShieldToken = GamblersLuckCapability.GamblersLuckShieldToken;

/** A gamblers-luck shield token feature (Gamblers' Luck promo). */
export class GamblersLuckShield extends TileFeature {
  static readonly simpleName = "GamblersLuckShield";

  private readonly shieldToken: GamblersLuckShieldToken | null;

  constructor(loc: Location);
  constructor(places: List<FeaturePointer>, shieldToken: GamblersLuckShieldToken | null);
  constructor(a: Location | List<FeaturePointer>, shieldToken?: GamblersLuckShieldToken | null) {
    if (a instanceof List) {
      super(a);
      this.shieldToken = shieldToken ?? null;
    } else {
      super(List.of(new FeaturePointer(Position.ZERO, GamblersLuckShield, a)));
      this.shieldToken = null;
    }
  }

  placeOnBoard(pos: Position, rot: Rotation): GamblersLuckShield {
    return new GamblersLuckShield(this.placeOnBoardPlaces(pos, rot), this.shieldToken);
  }

  setShieldToken(shieldToken: GamblersLuckShieldToken | null): GamblersLuckShield {
    return new GamblersLuckShield(this.places, shieldToken);
  }

  getShieldToken(): GamblersLuckShieldToken | null {
    return this.shieldToken;
  }
}
